package repositories: introduce path property for exact paths

cjp256 opened the pull request:
Paths pecifies an exact path relative to `url`.

In sources.list(5) manpage, an "exact path" is a suite that ends
with a "/", and cannot be combined with components.

Snapcraft will make this translation for apt, but exposes these
two distinct usages differently in snapcraft.yaml configuration.

- Add path property, enforcing dependencies & exclusivity in
  validate().  Make components and suites (and path) optional
  in the schema, but validate the various cases when instantiating.

- Refactor the unit tests to migrate to the current style & do
  a little organization to make them easier to follow.

- Add validate() for PackageRepositoryPpa as done previously with
  PackageRepositoryApt.

- Add apt_repo parameter for repo dependency injection. This will
  simplify test mocking while maintaining current behavior by
  specifying a default value of repo.Ubuntu.

- Add unit tests to cover new functionality.

- Minor specification tweaks/fixes:

  - Don't require the '/' if a path is specified, Snapcraft will
    just append it if needed).

  - Fix statement about path starting with "/", it should be
    "ending".

  - Replace usage of "absolute" path with "exact".

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

sergiusens commented on snapcraft/internal/meta/package_repository.py:
```diff
@@ -141,24 +152,40 @@ def __init__(
+            # Suites denoting exact path must end with '/'.
+            path = self.path
+            if not path.endswith("/"):
+                path = path + "/"
```
FWIW, this is also valid
```suggestion
                path += "/"
```

cjp256 replied:
Done.

sergiusens commented on snapcraft/internal/meta/package_repository.py:
```diff
@@ -120,13 +129,15 @@ def __init__(
+        path: Optional[str] = None,
+        suites: Optional[List[str]] = None,
         url: str,
+        apt_repo: Type[Ubuntu] = Ubuntu,
```
Is there a reason for this?

cjp256 replied:
There was for dependency injection, but I have since refactored install bits out of meta, so this will be dropped when I push.

sergiusens commented on snapcraft/internal/meta/package_repository.py:
```diff
@@ -141,24 +152,40 @@ def __init__(
         self.validate()
 
     def install(self, keys_path: Path) -> bool:
+        if not self.path and not self.components and not self.suites:
```
Can you add a code comment for this?

cjp256 replied:
Done.

sergiusens commented on snapcraft/internal/meta/package_repository.py:
```diff
@@ -141,24 +152,40 @@ def __init__(
+        elif self.suites:
+            suites = self.suites
+        else:
+            raise RuntimeError("no suites or path")
```
This would only happen if validation fails, correct?

cjp256 replied:
Yes, this should never happen.  I added a note (and another sanity check).

sergiusens reviewed (changes requested):
Pretty good, just some minor questions and code comment sprinkling

cjp256 commented:
I'll come back to this on Tuesday

cjp256 commented:
Should be good to go. I rewrote most of the error messages as well as improved some of the checks to make the outputs more precise.

And added validation/tests for architecture values.

sergiusens commented on snapcraft/internal/meta/package_repository.py:
```diff
@@ -181,40 +236,119 @@ def marshal(self) -> Dict[str, Any]:
-                resolution="You can specify a list of formats including 'deb' and/or 'deb-src'.",
-            )
+    def validate(self) -> None:  # noqa: C901
+        if self.architectures is not None:
```
valid architectures are dependant on base, core20 and i386 are invalid as is core18 or less with riscv

sergiusens replied:
This might be a fragile check, might be worth just letting apt fail and surface an error

cjp256 replied:
Good point.   It was mostly meant as a sanity check, riscv enablement would require it to be added here.

Longer term I'd hope for a shared deb arch enum that we could check against.

I can remove it if you want.

sergiusens replied:
All our enums are in ProjectOptions today, maybe add it there?

sergiusens replied:
Then again, it is not uncommon for an arch to be enabled post release, so this would require doing post facto work, removing might just be the better option if we can surface the error correctly

cjp256 replied:
Removed.

sergiusens reviewed (changes requested):
One last thing, looking good!

cjp256 commented:
Just the plainbox failures.
